File: walk_tracker.rb
# walk_tracker.rb
#!/usr/bin/env ruby
require 'json'
require 'date'

CONFIG_FILE = 'walk_config.json'
DATA_FILE = 'walks.json'

class WalkTracker
  attr_reader :weight, :walks

  def initialize
    @weight = 70.0
    @walks = []
    load_config
    load_walks
  end

  def load_config
    if File.exist?(CONFIG_FILE)
      config = JSON.parse(File.read(CONFIG_FILE))
      @weight = config['weight'] || 70.0
    end
  end

  def save_config
    File.write(CONFIG_FILE, JSON.pretty_generate({ 'weight' => @weight }))
  end

  def load_walks
    if File.exist?(DATA_FILE)
      @walks = JSON.parse(File.read(DATA_FILE))
    end
  end

  def save_walks
    File.write(DATA_FILE, JSON.pretty_generate(@walks))
  end

  def add_walk(distance, duration, weight = nil)
    weight ||= @weight
    speed = distance / (duration / 60.0)
    met = if speed < 3.0
            2.0
          elsif speed < 5.0
            3.0
          elsif speed < 6.5
            3.8
          elsif speed < 8.0
            5.0
          else
            6.0
          end
    calories = met * weight * (duration / 60.0)
    entry = {
      'date' => Time.now.iso8601,
      'distance' => distance,
      'duration' => duration,
      'speed' => speed,
      'weight' => weight,
      'calories' => calories.round(1)
    }
    @walks << entry
    save_walks
    puts "✅ Walk added: #{distance} km in #{duration} min, #{calories.round(1)} kcal"
  end

  def list_walks
    if @walks.empty?
      puts "No walks recorded."
      return
    end
    puts "\n📋 Walks:"
    @walks.each_with_index do |w, i|
      dt = w['date'][0..15].gsub('T', ' ')
      puts "#{i+1}. #{dt} | #{w['distance']} km | #{w['duration']} min | #{w['speed'].round(1)} km/h | #{w['calories']} kcal"
    end
  end

  def stats
    if @walks.empty?
      puts "No walks yet."
      return
    end
    total_dist = @walks.sum { |w| w['distance'] }
    total_dur = @walks.sum { |w| w['duration'] }
    total_cal = @walks.sum { |w| w['calories'] }
    puts "\n🚶 Walk Summary"
    puts "Total walks: #{@walks.size}"
    puts "Total distance: #{total_dist.round(2)} km"
    puts "Total duration: #{total_dur} min (#{(total_dur/60.0).round(1)} h)"
    puts "Total calories: #{total_cal.round(0)} kcal"
  end

  def export_csv(filename)
    require 'csv'
    CSV.open(filename, 'w') do |csv|
      csv << ["Date", "Distance (km)", "Duration (min)", "Speed (km/h)", "Weight (kg)", "Calories (kcal)"]
      @walks.each do |w|
        csv << [w['date'], w['distance'], w['duration'], w['speed'].round(1), w['weight'], w['calories']]
      end
    end
    puts "Exported #{@walks.size} walks to #{filename}"
  end
end

if ARGV.empty?
  puts "Usage: walk_tracker.rb [weight|add|list|stats|export]"
  exit
end

tracker = WalkTracker.new
cmd = ARGV.shift

case cmd
when 'weight'
  weight = ARGV.shift.to_f
  tracker.weight = weight
  tracker.save_config
  puts "Default weight set to #{weight} kg"
when 'add'
  if ARGV.size < 2
    puts "Usage: add <distance> <duration> [--weight <kg>]"
    exit
  end
  distance = ARGV.shift.to_f
  duration = ARGV.shift.to_f
  weight = nil
  if ARGV.include?('--weight')
    idx = ARGV.index('--weight')
    weight = ARGV[idx+1].to_f if idx
  end
  tracker.add_walk(distance, duration, weight)
when 'list'
  tracker.list_walks
when 'stats'
  tracker.stats
when 'export'
  filename = ARGV.shift || 'walks.csv'
  tracker.export_csv(filename)
else
  puts "Unknown command"
end

File: walk_tracker.py
# walk_tracker.py
import json
import os
from datetime import datetime
import argparse

CONFIG_FILE = "walk_config.json"
DATA_FILE = "walks.json"

class WalkTracker:
    def __init__(self):
        self.weight = 70.0  # default kg
        self.walks = []
        self.load_config()
        self.load_walks()

    def load_config(self):
        if os.path.exists(CONFIG_FILE):
            with open(CONFIG_FILE, "r") as f:
                config = json.load(f)
                self.weight = config.get("weight", 70.0)

    def save_config(self):
        with open(CONFIG_FILE, "w") as f:
            json.dump({"weight": self.weight}, f)

    def load_walks(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                self.walks = json.load(f)

    def save_walks(self):
        with open(DATA_FILE, "w") as f:
            json.dump(self.walks, f, indent=2)

    def add_walk(self, distance, duration, weight=None):
        if weight is None:
            weight = self.weight
        # Calculate speed km/h
        speed = distance / (duration / 60.0) if duration > 0 else 0
        # MET values based on speed (approx)
        if speed < 3.0:
            met = 2.0
        elif speed < 5.0:
            met = 3.0
        elif speed < 6.5:
            met = 3.8
        elif speed < 8.0:
            met = 5.0
        else:
            met = 6.0
        # Calories: MET * weight(kg) * duration(hours)
        calories = met * weight * (duration / 60.0)
        entry = {
            "date": datetime.now().isoformat(),
            "distance": distance,
            "duration": duration,
            "speed": speed,
            "weight": weight,
            "calories": round(calories, 1)
        }
        self.walks.append(entry)
        self.save_walks()
        print(f"✅ Walk added: {distance} km in {duration} min, {calories:.1f} kcal")

    def list_walks(self):
        if not self.walks:
            print("No walks recorded.")
            return
        print("\n📋 Walks:")
        for i, w in enumerate(self.walks, 1):
            dt = w["date"][:16].replace("T", " ")
            print(f"{i}. {dt} | {w['distance']:.2f} km | {w['duration']} min | {w['speed']:.1f} km/h | {w['calories']:.1f} kcal")

    def stats(self):
        if not self.walks:
            print("No walks yet.")
            return
        total_dist = sum(w["distance"] for w in self.walks)
        total_dur = sum(w["duration"] for w in self.walks)
        total_cal = sum(w["calories"] for w in self.walks)
        print("\n🚶 Walk Summary")
        print(f"Total walks: {len(self.walks)}")
        print(f"Total distance: {total_dist:.2f} km")
        print(f"Total duration: {total_dur} min ({total_dur/60:.1f} h)")
        print(f"Total calories: {total_cal:.0f} kcal")

    def export_csv(self, filename):
        import csv
        with open(filename, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(["Date", "Distance (km)", "Duration (min)", "Speed (km/h)", "Weight (kg)", "Calories (kcal)"])
            for w in self.walks:
                writer.writerow([w["date"], w["distance"], w["duration"], round(w["speed"], 1), w["weight"], w["calories"]])
        print(f"Exported {len(self.walks)} walks to {filename}")

def main():
    parser = argparse.ArgumentParser(description="Walk Tracker")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    weight_parser = subparsers.add_parser("weight")
    weight_parser.add_argument("kg", type=float)

    add_parser = subparsers.add_parser("add")
    add_parser.add_argument("distance", type=float)
    add_parser.add_argument("duration", type=float)
    add_parser.add_argument("--weight", type=float, help="Weight for this walk (kg)")

    subparsers.add_parser("list")
    subparsers.add_parser("stats")

    export_parser = subparsers.add_parser("export")
    export_parser.add_argument("filename", default="walks.csv", nargs="?")

    args = parser.parse_args()
    tracker = WalkTracker()

    if args.cmd == "weight":
        tracker.weight = args.kg
        tracker.save_config()
        print(f"Default weight set to {args.kg} kg")
    elif args.cmd == "add":
        tracker.add_walk(args.distance, args.duration, args.weight)
    elif args.cmd == "list":
        tracker.list_walks()
    elif args.cmd == "stats":
        tracker.stats()
    elif args.cmd == "export":
        tracker.export_csv(args.filename)

if __name__ == "__main__":
    main()
